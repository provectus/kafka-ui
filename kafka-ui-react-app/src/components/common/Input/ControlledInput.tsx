import React from 'react';
import { Controller } from 'react-hook-form';

import Input, { InputProps } from './Input';

const ControlledInput = ({
  name,
  control,
  ...restProps
}: // eslint-disable-next-line @typescript-eslint/no-explicit-any
InputProps & { control: any }) => {
  return (
    <Controller
      control={control}
      name={name ?? ''}
      render={({ field }) => <Input {...restProps} {...field} />}
    />
  );
};

export default ControlledInput;
