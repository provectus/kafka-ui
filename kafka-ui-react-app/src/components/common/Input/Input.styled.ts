import styled, { css } from 'styled-components';

export interface InputProps {
  inputSize?: 'S' | 'M' | 'L';
  search: boolean;
}

const INPUT_SIZES = {
  S: '24px',
  M: '32px',
  L: '40px',
};

export const Wrapper = styled.div`
  position: relative;
  &:hover {
    svg:first-child {
      fill: ${({ theme }) => theme.input.icon.hover};
    }
  }
  svg:first-child {
    position: absolute;
    top: 8px;
    line-height: 0;
    z-index: 1;
    left: 12px;
    right: unset;
    height: 16px;
    width: 16px;
    fill: ${({ theme }) => theme.input.icon.color};
  }
`;

export const Input = styled.input<InputProps>(
  ({ theme: { input }, inputSize, search }) => css`
    background-color: ${input.backgroundColor.normal};
    border: 1px ${input.borderColor.normal} solid;
    border-radius: 4px;
    color: ${input.color.normal};
    height: ${inputSize && INPUT_SIZES[inputSize]
      ? INPUT_SIZES[inputSize]
      : '40px'};
    width: 100%;
    padding-left: ${search ? '36px' : '12px'};
    font-size: 14px;

    &::placeholder {
      color: ${input.color.placeholder.normal};
      font-size: 14px;
    }
    &:hover {
      border-color: ${input.borderColor.hover};
    }
    &:focus {
      outline: none;
      border-color: ${input.borderColor.focus};
      &::placeholder {
        color: transparent;
      }
    }
    &:disabled {
      color: ${input.color.disabled};
      border-color: ${input.borderColor.disabled};
      background-color: ${input.backgroundColor.disabled};
      cursor: not-allowed;
    }
    &:read-only {
      color: ${input.color.readOnly};
      border: none;
      background-color: ${input.backgroundColor.readOnly};
      &:focus {
        &::placeholder {
          color: ${input.color.placeholder.readOnly};
        }
      }
      cursor: not-allowed;
    }
  `
);

export const FormError = styled.p`
  color: ${({ theme }) => theme.input.error};
  font-size: 12px;
`;

export const InputHint = styled.p`
  font-size: 0.85rem;
  margin-top: 0.25rem;
  color: ${({ theme }) => theme.clusterConfigForm.inputHintText.secondary};
`;
