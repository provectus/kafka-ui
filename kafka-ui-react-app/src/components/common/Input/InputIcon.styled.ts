import styled from 'styled-components';
import { Colors } from 'theme/theme';

interface Props {
  className: string;
  position: 'left' | 'right';
  inputSize: 'M' | 'L';
}

const StyledIcon = styled.i<Props>`
  position: absolute;
  top: 50%;
  line-height: 0;
  z-index: 1;
  left: ${(props) => (props.position === 'left' ? '12px' : 'unset')};
  right: ${(props) => (props.position === 'right' ? '15px' : 'unset')};
  height: 11px;
  width: 11px;
  color: ${Colors.neutral[70]};
`;

export default StyledIcon;
