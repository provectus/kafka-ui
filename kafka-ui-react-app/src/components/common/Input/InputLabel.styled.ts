import styled from 'styled-components';

export const InputLabel = styled.label`
  font-weight: 500;
  font-size: 12px;
  line-height: 20px;
  color: ${({ theme }) => theme.input.label.color};
  input[type='checkbox'] {
    display: inline-block;
    margin-right: 8px;
    vertical-align: text-top;
  }
`;
