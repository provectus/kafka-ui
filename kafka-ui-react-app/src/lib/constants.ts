import { SelectOption } from 'components/common/Select/Select';
import { ConfigurationParameters } from 'generated-sources';

declare global {
  interface Window {
    basePath: string;
  }
}

export const BASE_PARAMS: ConfigurationParameters = {
  basePath: window.basePath || '',
  credentials: 'include',
  headers: {
    'Content-Type': 'application/json',
  },
};

export const TOPIC_NAME_VALIDATION_PATTERN = /^[a-zA-Z0-9._-]+$/;
export const SCHEMA_NAME_VALIDATION_PATTERN = /^[.,A-Za-z0-9_/-]+$/;

export const TOPIC_CUSTOM_PARAMS_PREFIX = 'customParams';
export const TOPIC_CUSTOM_PARAMS: Record<string, string> = {
  'compression.type': 'producer',
  'leader.replication.throttled.replicas': '',
  'message.downconversion.enable': 'true',
  'segment.jitter.ms': '0',
  'flush.ms': '9223372036854775807',
  'follower.replication.throttled.replicas': '',
  'segment.bytes': '1073741824',
  'flush.messages': '9223372036854775807',
  'message.format.version': '2.3-IV1',
  'file.delete.delay.ms': '60000',
  'max.compaction.lag.ms': '9223372036854775807',
  'min.compaction.lag.ms': '0',
  'message.timestamp.type': 'CreateTime',
  preallocate: 'false',
  'min.cleanable.dirty.ratio': '0.5',
  'index.interval.bytes': '4096',
  'unclean.leader.election.enable': 'true',
  'retention.bytes': '-1',
  'delete.retention.ms': '86400000',
  'segment.ms': '604800000',
  'message.timestamp.difference.max.ms': '9223372036854775807',
  'segment.index.bytes': '10485760',
};

export const MILLISECONDS_IN_WEEK = 604_800_000;
export const MILLISECONDS_IN_DAY = 86_400_000;
export const MILLISECONDS_IN_SECOND = 1_000;

export const NOT_SET = -1;
export const BYTES_IN_GB = 1_073_741_824;
export const BUILD_VERSION_PATTERN = /v\d.\d.\d/;

export const PER_PAGE = 25;
export const MESSAGES_PER_PAGE = '100';

export const GIT_REPO_LINK = 'https://github.com/provectus/kafka-ui';
export const GIT_REPO_LATEST_RELEASE_LINK =
  'https://api.github.com/repos/provectus/kafka-ui/releases/latest';

export const LOCAL_STORAGE_KEY_PREFIX = 'kafka-ui';

export enum AsyncRequestStatus {
  initial = 'initial',
  pending = 'pending',
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export const QUERY_REFETCH_OFF_OPTIONS = {
  refetchOnMount: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
};

// Cluster Form Constants
export const AUTH_OPTIONS: SelectOption[] = [
  { value: 'SASL/JAAS', label: 'SASL/JAAS' },
  { value: 'SASL/GSSAPI', label: 'SASL/GSSAPI' },
  { value: 'SASL/OAUTHBEARER', label: 'SASL/OAUTHBEARER' },
  { value: 'SASL/PLAIN', label: 'SASL/PLAIN' },
  { value: 'SASL/SCRAM-256', label: 'SASL/SCRAM-256' },
  { value: 'SASL/SCRAM-512', label: 'SASL/SCRAM-512' },
  { value: 'Delegation tokens', label: 'Delegation tokens' },
  { value: 'SASL/LDAP', label: 'SASL/LDAP' },
  { value: 'SASL/AWS IAM', label: 'SASL/AWS IAM' },
  { value: 'mTLS', label: 'mTLS' },
];

export const SECURITY_PROTOCOL_OPTIONS: SelectOption[] = [
  { value: 'SASL_SSL', label: 'SASL_SSL' },
  { value: 'SASL_PLAINTEXT', label: 'SASL_PLAINTEXT' },
];
export const METRICS_OPTIONS: SelectOption[] = [
  { value: 'JMX', label: 'JMX' },
  { value: 'PROMETHEUS', label: 'PROMETHEUS' },
];
