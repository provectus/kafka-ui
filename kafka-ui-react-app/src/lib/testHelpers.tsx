import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
