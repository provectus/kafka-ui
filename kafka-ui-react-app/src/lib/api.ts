import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
