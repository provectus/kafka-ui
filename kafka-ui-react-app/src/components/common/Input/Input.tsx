import React from 'react';
import { useFormContext } from 'react-hook-form';
import { styled } from 'lib/themedStyles';

import StyledIcon from './InputIcon.styled';
import StyledInput, { StyledInputProps } from './Input.styled';

export interface InputProps
  extends React.InputHTMLAttributes<HTMLInputElement>,
    Omit<StyledInputProps, 'hasLeftIcon'> {
  name?: string;
  leftIcon?: string;
  rightIcon?: string;
}

const Input: React.FC<InputProps> = ({
  className,
  name,
  leftIcon,
  rightIcon,
  inputSize = 'L',
  ...rest
}) => {
  const methods = useFormContext();
  return (
    <div className={className}>
      {leftIcon && (
        <StyledIcon
          className={leftIcon}
          position="left"
          inputSize={inputSize}
        />
      )}
      {name ? (
        <StyledInput
          className={className}
          inputSize={inputSize}
          {...methods.register(name)}
          hasLeftIcon={!!leftIcon}
          {...rest}
        />
      ) : (
        <StyledInput
          className={className}
          inputSize={inputSize}
          hasLeftIcon={!!leftIcon}
          {...rest}
        />
      )}
      {rightIcon && (
        <StyledIcon
          className={rightIcon}
          position="right"
          inputSize={inputSize}
        />
      )}
    </div>
  );
};

const InputWrapper = styled(Input)`
  position: relative;
`;

export default InputWrapper;
