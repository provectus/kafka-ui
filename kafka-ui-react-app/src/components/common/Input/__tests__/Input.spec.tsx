import Input, { InputProps } from 'components/common/Input/Input';
import React from 'react';
import { screen } from '@testing-library/react';
import { render } from 'lib/testHelpers';
import userEvent from '@testing-library/user-event';

// Mock useFormContext
let component: HTMLInputElement;

const setupWrapper = (props?: Partial<InputProps>) => (
  <Input name="test" {...props} />
);
jest.mock('react-hook-form', () => ({
  useFormContext: () => ({
    register: jest.fn(),

    // Mock methods.getValues and methods.setValue
    getValues: jest.fn(() => {
      return component.value;
    }),
    setValue: jest.fn((key, val) => {
      component.value = val;
    }),
  }),
}));

describe('Custom Input', () => {
  describe('with no icons', () => {
    const getInput = () => screen.getByRole('textbox');

    it('to be in the document', () => {
      render(setupWrapper());
      expect(getInput()).toBeInTheDocument();
    });
  });
  describe('number', () => {
    const getInput = () => screen.getByRole<HTMLInputElement>('spinbutton');

    describe('input', () => {
      it('allows user to type numbers only', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, 'abc131');
        expect(input).toHaveValue(131);
      });

      it('allows user to type negative values', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '-2');
        expect(input).toHaveValue(-2);
      });

      it('allows user to type positive values only', async () => {
        render(setupWrapper({ type: 'number', positiveOnly: true }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '-2');
        expect(input).toHaveValue(2);
      });

      it('allows user to type decimal', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '2.3');
        expect(input).toHaveValue(2.3);
      });

      it('allows user to type integer only', async () => {
        render(setupWrapper({ type: 'number', integerOnly: true }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '2.3');
        expect(input).toHaveValue(23);
      });

      it("not allow '-' appear at any position of the string except the start", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '2-3');
        expect(input).toHaveValue(23);
      });

      it("not allow '.' appear at the start of the string", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '.33');
        expect(input).toHaveValue(33);
      });

      it("not allow '.' appear twice in the string", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.type(input, '3.3.3');
        expect(input).toHaveValue(3.33);
      });
    });

    describe('paste', () => {
      it('allows user to paste numbers only', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('abc131');
        expect(input).toHaveValue(131);
      });

      it('allows user to paste negative values', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('-2');
        expect(input).toHaveValue(-2);
      });

      it('allows user to paste positive values only', async () => {
        render(setupWrapper({ type: 'number', positiveOnly: true }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('-2');
        expect(input).toHaveValue(2);
      });

      it('allows user to paste decimal', async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('2.3');
        expect(input).toHaveValue(2.3);
      });

      it('allows user to paste integer only', async () => {
        render(setupWrapper({ type: 'number', integerOnly: true }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('2.3');
        expect(input).toHaveValue(23);
      });

      it("not allow '-' appear at any position of the pasted string except the start", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('2-3');
        expect(input).toHaveValue(23);
      });

      it("not allow '.' appear at the start of the pasted string", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('.33');
        expect(input).toHaveValue(0.33);
      });

      it("not allow '.' appear twice in the pasted string", async () => {
        render(setupWrapper({ type: 'number' }));
        const input = getInput();
        component = input;
        await userEvent.click(input);
        await userEvent.paste('3.3.3');
        expect(input).toHaveValue(3.33);
      });
    });
  });
});
