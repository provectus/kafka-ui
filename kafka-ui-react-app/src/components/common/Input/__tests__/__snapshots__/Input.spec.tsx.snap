// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Input with icons matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-dlfnuX gxtwfA"
    >
      <i
        class="sc-bdfBQB gNwdVP fas fa-address-book"
      />
      <input
        class="sc-gsTEea eQzZSW sc-dlfnuX gxtwfA"
      />
      <i
        class="sc-bdfBQB gJpiMQ fas fa-address-book"
      />
    </div>
  </div>
</body>
`;

exports[`Custom Input with no icons matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-dlfnuX gxtwfA"
    >
      <input
        class="sc-gsTEea QNjxA sc-dlfnuX gxtwfA"
      />
    </div>
  </div>
</body>
`;
